include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -191,13 +205,124 @@ class cuda_vm_resource : public memory_resource<memory_kind::device>,
       stat_.allocated_blocks -= blocks_to_free.size();
     }
     blocks_to_free.clear();  // free the physical memory
+    if (release_va)
+      va_freed = release_unused_va();
+    return { mem_freed, va_freed };
   }
 
  protected:
   void do_deallocate(void *ptr, size_t size, size_t alignment) override {
     deallocate_impl(ptr, size, alignment, true);
   }
 
+  int find_va_region(CUdeviceptr ptr) {
+    for (size_t i = 0; i < va_regions_.size(); i++)
+      if (va_regions_[i].address_range.contains(ptr))
+        return i;
+    return -1;
+  }
+
+  size_t release_unused_va() {
+    struct RangeDesc {
+      int range_idx, region_idx;
+      int start_block, end_block;
+    };
+    std::vector<cuvm::CUMem> blocks_to_free;
+    std::vector<cuvm::CUMemAddressRange> va_ranges_to_free;
+    std::vector<RangeDesc> ranges_to_free;
+    blocks_to_free.reserve(stat_.allocated_blocks);
+    va_ranges_to_free.reserve(va_ranges_.size());
+    ranges_to_free.reserve(va_ranges_.size());
+    size_t va_freed = 0;
+
+    {
+      std::unique_lock pool_guard(pool_lock_, std::try_to_lock);
+      std::unique_lock mem_guard(mem_lock_, std::try_to_lock);
+
+      // the whole block below musn't throw or it'll leave the resource in an inconsistent state
+      [&]() noexcept {
+        for (int i = 0, n = va_ranges_.size(); i < n; i++) {
+          auto start = va_ranges_[i].ptr();
+          auto end = va_ranges_[i].end();
+          if (free_va_.contains(reinterpret_cast<void*>(start), reinterpret_cast<void*>(end))) {
+            int region_idx = find_va_region(start);
+            assert(region_idx >= 0);
+            va_region &region = va_regions_[region_idx];
+
+            ptrdiff_t offset = region.address_range.ptr() - start;
+            assert(offset % block_size_ == 0);
+            int num_blocks = (end - start) / block_size_;
+            int start_block = offset / block_size_;  // index of the first block
+            int end_block = start_block + num_blocks;  // index of the last block
+
+            int avail = region.available.find(true, start_block);
+            while (avail < end_block) {
+              int end_avail = std::min<int>(region.available.find(false, avail + 1), end_block);
+              for (int b = avail; b < end_avail; b++) {
+                blocks_to_free.push_back(region.unmap_block(b));
+                stat_.total_unmaps++;
+              }
+              auto *start_ptr = region.block_ptr<char>(avail);
+              auto *end_ptr = region.block_ptr<char>(end_avail);
+              free_mapped_.get_specific_block(start_ptr, end_ptr);
+
+              if (end_avail == end_block)
+                break;
```

Comment:
Wouldn't this be covered by L259?

Reply:
If we already reach `end_block` there's no point in running the search in L271.